### PR TITLE
Profiling

### PR DESCRIPTION
**Pull request by MaksymilianDemitraszek**

### CODE REVIEW

**Comment by MaksymilianDemitraszek** [comment]
This includes core of the profiler, docs are going to be included in a separate PR

File: protostar/profiler/contract_profiler.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+    def build_memhole_samples(
+        self,
+        instructions: List[Instruction],
+        tracer_data: TracerData,
+        accessed_memory: Set[RelocatableValue],
+        segments: MemorySegmentManager,
+        segment_offsets,
+    ) -> List[Sample]:
+        # Address -> Pc of instruction which accessed the address last
+        accessed_by: Dict[Address, Address] = {}
```
Comment (kasperski95) [inline protostar/profiler/contract_profiler.py:269]: I am confused by `Pc of instruction` and `PC` in general. I'd use `InstructionAddress` instead of PC (Program Counter) and renamed the `Address` type to `MemoryAddress`.

File: tests/helper.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import sys
```
Comment (kasperski95) [inline tests/helper.py:1]: Is this class used somewhere? Maybe it can be moved to conftest file.

File: protostar/profiler/contract_profiler.py [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+Address = int
+FunctionID = str
+
+
+@dataclass(frozen=True)
+class Function:
+    """
+    Represents a function in a contract
+
+    id -- uniuque name of a function
```
Comment (kasperski95) [inline protostar/profiler/contract_profiler.py:27]:
```suggestion
    id -- unique name of a function
```

File: protostar/profiler/contract_profiler.py [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+    filename: str
+    start_line: int
+
+
+@dataclass(frozen=True)
+class Instruction:
+    """
+    Represents a instruction under certain pc in a contract
+
+    id -- uniuque id of a instruction
```
Comment (kasperski95) [inline protostar/profiler/contract_profiler.py:42]:
```suggestion
    id -- unique id of a instruction
```

File: protostar/profiler/contract_profiler.py [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+    pc -- pc instruction is placed under
+    function -- function from the instuction has been generated
```
Comment (kasperski95) [inline protostar/profiler/contract_profiler.py:44]:
```suggestion
    function -- function from the instruction has been generated
```

File: protostar/profiler/contract_profiler.py [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+        for func in functions:
+            if func.id == function_id:
+                return func
+        assert False
+
+    def create_instruction_list(
+        self, functions: List[Function], tracer_data: TracerData
+    ) -> List[Instruction]:
+        """
+        Builds a list of instructions in the contract and assings them to functions they come from
```
Comment (kasperski95) [inline protostar/profiler/contract_profiler.py:150]:
```suggestion
        Builds a list of instructions in the contract and assigns them to functions they come from
```

File: protostar/profiler/contract_profiler.py [hunk 7]
```diff
@@ -0,0 +1,319 @@
+# pylint: disable=invalid-name
```
Comment (kasperski95) [inline protostar/profiler/contract_profiler.py:1]: Consider creating `README.md` in the profiler module with high level overview.

**Review by kasperski95 — APPROVED** [review]
LGTM

File: protostar/commands/test/test_command.py [hunk 8]
```diff
@@ -101,6 +102,13 @@ def arguments(self) -> List[Command.Argument]:
                 ),
                 type="bool",
             ),
+            Command.Argument(
+                name="profiling",
+                description=(
+                    "Run profiling for a test contract. Works only for a single test case."
```
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/commands/test/test_command.py:108]: It will be included in docs, this is too complicated to do it in cli.

[the remark this answers is not in the document]
Comment (kasperski95) [reply protostar/commands/test/test_command.py:108]: "Protostar will generate a file that can be opened with https://github.com/google/pprof"

File: protostar/profiler/contract_profiler.py (hunk 1 above)
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py:269]: > MemoryAddress
I do not agree, it is unnecessary wordy

>InstructionAddress
I can consider this in few places

File: tests/helper.py (hunk 2 above)
Comment (MaksymilianDemitraszek) [reply tests/helper.py:1]: Nah, accidental push

File: protostar/profiler/contract_profiler.py (hunk 7 above)
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py:1]: #983

Comment (mkaput) [inline protostar/profiler/contract_profiler.py:1]: nitpick: what name is invalid? could we avoid global linter disable comment here?

File: tests/e2e/test_testing.py [hunk 9]
```diff
@@ -19,6 +20,25 @@ def test_safe_collecting(protostar):
...
+
+
+@pytest.mark.usefixtures("init")
+def test_profile_fuzz(protostar, copy_fixture):
+    copy_fixture("fuzz_test.cairo", "./tests")
+    result = protostar(
+        ["test", "--profiling", "tests/fuzz_test.cairo"], ignore_exit_code=True
+    )
+    assert "Fuzz tests cannot be profiled" in result
+    assert "profile.pb.gz" not in listdir(".")
```
Comment (mkaput) [inline tests/e2e/test_testing.py:39]: nitpick
```suggestion
    assert not Path("profile.pb.gz").exists()
```

File: protostar/testing/test_runner.py [hunk 10]
```diff
@@ -177,12 +181,14 @@ async def _invoke_test_cases(
...
-            test_result = await self._invoke_test_case(test_case, execution_state)
+            test_result = await self._invoke_test_case(
+                test_case, execution_state, profiling=self.profiling
+            )
             self.shared_tests_state.put_result(test_result)
 
     @staticmethod
     async def _invoke_test_case(
-        test_case: TestCase, initial_state: TestExecutionState
+        test_case: TestCase, initial_state: TestExecutionState, profiling=False
```
Comment (mkaput) [inline protostar/testing/test_runner.py:191]: I am thinking if it wouldn't make more sense to put this flag inside `TestConfig`. Much less passing via args and it would be possible to enable via cheatcodes in setup.

File: protostar/testing/starkware/test_execution_state.py [hunk 11]
```diff
@@ -43,6 +43,7 @@ async def from_test_suite_definition(
         starknet.cheatable_state.cheatable_state.contract_address_to_class_hash_map[
             contract.contract_address
         ] = 0
+
```
Comment (mkaput) [inline protostar/testing/starkware/test_execution_state.py:46]: unnecessary change outside scope of this PR

File: protostar/starknet/cheatable_execute_entry_point.py [hunk 12]
```diff
@@ -27,23 +29,60 @@
...
+# TODO(maksymiliandemitraszek): Enable it again
+# pylint: disable=too-many-branches
 class CheatableExecuteEntryPoint(ExecuteEntryPoint):
     cheatcode_factory: Optional["CheatcodeFactory"] = None
+    samples: List[ContractProfile] = []
+    contract_callstack: List[str] = []
+    profiling = False
 
-    def _run(
+    def _run(  # type: ignore
```
Comment (mkaput) [inline protostar/starknet/cheatable_execute_entry_point.py:85]: why?

File: protostar/protostar_cli.py [hunk 13]
```diff
@@ -39,7 +39,6 @@ def __init__(
         self._version_manager = version_manager
         self._start_time = start_time
         self._project_cairo_path_builder = project_cairo_path_builder
-
```
Comment (mkaput) [inline protostar/protostar_cli.py:42]: Unnecessary change outside scope of this PR

File: protostar/profiler/contract_profiler.py (hunk 1 above)
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py]: Gave it an afterthough: instruction address is a bit better, but pc is used as an internal convention in cairo compiler/vm and i prefer to be consistent as this code touches those parts heavily

File: protostar/starknet/cheatable_execute_entry_point.py [hunk 14]
```diff
@@ -1,5 +1,7 @@
+from dataclasses import dataclass
 import logging
-from typing import TYPE_CHECKING, Optional, Tuple, cast, Any
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, List, Optional, Tuple, cast
```
Comment (mkaput) [inline protostar/starknet/cheatable_execute_entry_point.py:4]:
```suggestion
from typing import TYPE_CHECKING, Any, Optional, cast
```

nitpick, we have Python 3.9 available

**Review by mkaput — CHANGES_REQUESTED** [review]

**Review by mkaput — CHANGES_REQUESTED** [review]

File: protostar/starknet/cheatable_execute_entry_point.py (hunk 12 above)
Comment (MaksymilianDemitraszek) [reply protostar/starknet/cheatable_execute_entry_point.py:85]: Probably a leftover

File: protostar/profiler/contract_profiler.py (hunk 7 above)
Comment (mkaput) [reply protostar/profiler/contract_profiler.py:1]: f*ck this lint, can we disable it globally?

File: protostar/profiler/transaction_profiler.py [hunk 15]
```diff
@@ -0,0 +1,211 @@
+# pylint: disable=invalid-name
+from dataclasses import dataclass, replace
+import itertools
+from typing import TYPE_CHECKING, Tuple
+
+from protostar.profiler.contract_profiler import (
```
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/profiler/transaction_profiler.py:6]: it is only the function parameter type that is passed to `merge_profiles`, we could move it to `transaction_profiler` but it believe it belongs to cheatable exacute entry point

File: protostar/profiler/contract_profiler.py [hunk 16]
```diff
@@ -0,0 +1,317 @@
...
+        for trace_entry, mem_acc in zip(tracer_data.trace, tracer_data.memory_accesses):
+            frame_pcs = self.get_callstack(fp=trace_entry.fp, pc=trace_entry.pc)
+            addresses: list[Address] = [mem_acc[d] for d in ["dst", "op0", "op1"]]
+            for addr in addresses:
+                # Casting to Addres because adresses have been already relocated
+                # TODO(maksymiliandemitraszek) check if that's correct
+                accessed_by[addr] = cast(Address, trace_entry.pc)
+                pc_to_callstack[trace_entry.pc] = frame_pcs
+
+        blame_pc = functools.partial(self.blame_pc, accessed_by)
```
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py:267]: I really wanted to use partial

File: protostar/profiler/contract_profiler.py [hunk 17]
```diff
@@ -0,0 +1,317 @@
...
+            pc = blame_pc(address)
+            callstack = [
+                self.find_instruction(instructions, frame_pc)
+                for frame_pc in pc_to_callstack[pc]
+            ]
+            samples.append(Sample(value=1, callstack=callstack))
+        return samples
+
+    def get_callstack(self, fp: Address, pc: Address) -> list[Address]:
+        frame_pcs = [pc]
```
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py:283]: I am not sure if I see it, we only append pc to the list here

File: protostar/profiler/contract_profiler.py [hunk 18]
```diff
@@ -0,0 +1,317 @@
...
+    def __init__(self, initial_fp: Address, memory: MemoryDict):
+        self._initial_fp = initial_fp
+        self._memory = memory
+
+    def collect_contract_functions(self, tracer_data: TracerData) -> list[Function]:
+        identifiers_dict = tracer_data.program.identifiers.as_dict()
+        assert tracer_data.program.debug_info is not None
+        is_label: Callable[[IdentifierDefinition], bool] = lambda ident: isinstance(
+            ident, LabelDefinition
+        )
```
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py:93]: Inlining, it is a leftover from old code, it had more sense at the time

[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py:93]: (it was `is_instruction` and was there to document more complex condition)

File: protostar/profiler/contract_profiler.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (mkaput) [reply protostar/profiler/contract_profiler.py:268]: do as you wish but really it's harder to read this way

File: protostar/profiler/contract_profiler.py [hunk 19]
```diff
@@ -0,0 +1,317 @@
...
+    memhole_samples: list[Sample]
+    contract_call_callstacks: list[list[Instruction]]
+
+
+class ProfilerContext:
+    def __init__(self, initial_fp: Address, memory: MemoryDict):
+        self._initial_fp = initial_fp
+        self._memory = memory
+
+    def collect_contract_functions(self, tracer_data: TracerData) -> list[Function]:
```
[the remark this answers is not in the document]
Comment (mkaput) [reply protostar/profiler/contract_profiler.py:87]: right now this is more a bag functions attached to state, which do not touch the state at all. that's how I see it

[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py:87]: Alternatively we could pass memory to every function

[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py:87]: Discussed on slack

File: protostar/profiler/contract_profiler.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py:268]: changed

File: protostar/profiler/contract_profiler.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py:284]: Added docs

File: protostar/profiler/contract_profiler.py [hunk 20]
```diff
@@ -0,0 +1,317 @@
...
+            samples.append(Sample(value=1, callstack=callstack))
+        return samples
+
+    def get_callstack(self, fp: Address, pc: Address) -> list[Address]:
+        frame_pcs = [pc]
+        while fp > self._initial_fp:
+            fp_val = self._memory[fp - 2]
+            pc_val = self._memory[fp - 1]
+            assert isinstance(fp_val, Address)
+            assert isinstance(pc_val, Address)
```
[the remark this answers is not in the document]
Comment (MaksymilianDemitraszek) [reply protostar/profiler/contract_profiler.py:289]: Typewise - not, runtimewise - i think so

File: protostar/testing/test_runner.py (hunk 10 above)
Comment (MaksymilianDemitraszek) [reply protostar/testing/test_runner.py:191]: Yess, maybe a good idea would be moving other flags there as well? e.g. `disable_hint_validation_in_user_contracts`

File: protostar/profiler/contract_profiler.py [hunk 21]
```diff
@@ -55,11 +50,11 @@ class Instruction:
 class Sample:
     """
     Sample of a resource
-
-    value -- value of a sample (for example 1 memory hole)
-    callstack -- instruction callstack for the reported sample
-    Each instruction has an function assigned so it is easy to deduce the
-    calltree of an functions (pprof format requires samples to be assigned to instruction not functions)
+    Attributes:
```
Comment (mkaput) [inline protostar/profiler/contract_profiler.py:53]: FYI in next PR, you can put docstring for each field after the field. This should be picked by IDEs for example.

```python
class Foo:
    """Class docs"""

    field: int
    """Field docs"""

    moo: bool
    """moo docs"""

File: protostar/profiler/pprof.py [hunk 22]
```diff
@@ -6,33 +6,40 @@
...
+            string_table.append(val)
+        return string_ids[val]
+
 
 
 def to_protobuf(profile_obj: TransactionProfile) -> Profile:
     profile = Profile()
+    id_generator = StringIDGenerator()
     string_table = [""]
     string_ids = {"": 0}
```
Comment (mkaput) [inline protostar/profiler/pprof.py:26]:
```suggestion
```

File: website/docs/cli-reference.md [hunk 23]
```diff
@@ -205,7 +205,7 @@ Only re-run failed and broken test cases.
 #### `--no-progress-bar`
 Disable progress bar.
 #### `--profiling`
-Run profiling for a test contract. Works only for a single test case.
+Run profiling for a test contract. Works only for a single test case.Protostar generates a file that can be opened with https://github.com/google/pprof
```
Comment (mkaput) [inline website/docs/cli-reference.md:208]:
```suggestion
Run profiling for a test contract. Works only for a single test case. Protostar generates a file that can be opened with [pprof](https://github.com/google/pprof).
```

File: protostar/protostar_cli.py (hunk 13 above)
Comment (mkaput) [reply protostar/protostar_cli.py:42]: The change is still there ;)

File: protostar/testing/test_case_runners/test_case_runner_factory.py [hunk 24]
```diff
@@ -21,11 +21,10 @@ def make(self, test_case: TestCase) -> TestCaseRunner:
         mode = self._state.config.mode
 
         assert mode, "Test mode should be determined at this point."
-
```
Comment (mkaput) [inline protostar/testing/test_case_runners/test_case_runner_factory.py:24]: nitpick: this file behaviour is not changed at all here, consider reverting changes

File: protostar/testing/test_runner.py [hunk 25]
```diff
@@ -193,7 +198,9 @@ async def _invoke_test_cases(
         execution_state: TestExecutionState,
     ) -> None:
         for test_case in test_suite.test_cases:
-            test_result = await self._invoke_test_case(test_case, execution_state)
+            test_result = await self._invoke_test_case(
+                test_case, execution_state
+            )
```
Comment (mkaput) [inline protostar/testing/test_runner.py:203]:
```suggestion
            test_result = await self._invoke_test_case(test_case, execution_state)
```

nitpick: unnecessary change

**Review by mkaput — APPROVED** [review]

File: protostar/protostar_cli.py (hunk 13 above)
Comment (MaksymilianDemitraszek) [reply protostar/protostar_cli.py:42]: Oh there are two of them
